File: Data Structures & Algorithms/combinations/submission-0.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        def solve(i, temp):
            if len(temp) == k:
                res.append(temp.copy())
                return
            if i > n:
                return
            temp.append(i)
            solve(i+1, temp)
            temp.pop()
            solve(i+1, temp)
        solve(1, [])
        return res

File: Data Structures & Algorithms/generate-parentheses/submission-0.cpp
class Solution {
public:
    void solve(int n, int opened, int closed, string s, vector<string> &res){
        if(closed == opened && opened == n){
            res.push_back(s);
            return;
        }
        if(opened < n) solve(n, opened+1, closed, s + '(', res);
        if(closed < opened) solve(n, opened, closed + 1, s + ')', res);
    }
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        solve(n, 0, 0, "", res);
        return res;
    }
};
